### PR TITLE
Pttb 245 system should only accept new password

### PR DESCRIPTION
**Pull request by piyushmhrzn**
1. Used try-catch method for forgetPass() and resetPass() function.
2. Implemented Logger for all Exceptions.
3. Used local scope in User Model to get user by their email.
4. Created checkOldPass() function in UserService which will check if the new password is same as the old password
5. Dependency Injected for UserService and AuthController
6. Doc Blocks implemented for comments
7. Implemented HttpResponse trait for uniform messages.

### CODE REVIEW

**Review by sahajmalla — CHANGES_REQUESTED** [review]
If the controller has other than these resourceful functions, use actions class.

**Comment by piyushmhrzn** [comment]
Used single action controllers for forgot password and reset password actions

File: app/Services/UserService.php [hunk 1]
```diff
@@ -65,14 +113,39 @@ public static function roles()
...
         return true;
     }
 
-    public static function resetPassword(array $validatedResetPass): bool
+    /**
+     *
+     * @param array $validatedResetPass
+     * @return boolean
+     */
+    public function checkOldPass(array $validatedResetPass): bool
```
Comment (sahajmalla) [inline app/Services/UserService.php:133]: do this in form request class

**Review by sahajmalla — CHANGES_REQUESTED** [review]
Do all validation in form request class

File: app/Http/Controllers/Actions/Auth/ResetPasswordAction.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @throws ModelNotFoundException
+     * @throws QueryException
+     */
+    public function __invoke(PasswordResetRequest $request): JsonResponse
+    {
+        $validatedResetPass = $request->validated();
+        try {
+            // Check if the old pasword matches new password
+            $checkOldPass = $this->userService->checkOldPass($validatedResetPass);
+            if (!$checkOldPass){
```
[the remark this answers is not in the document]
Comment (piyushmhrzn) [reply app/Http/Controllers/Actions/Auth/ResetPasswordAction.php:46]: ok

File: app/Services/UserService.php (hunk 1 above)
Comment (piyushmhrzn) [reply app/Services/UserService.php:133]: ok

File: app/Http/Requests/ForgotPasswordRequest.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        return [
+            'email' => 'required|email|max:255|exists:users,email'
+        ];
+    }
+
+    public function messages()
+    {
+        return [
+            'email.exists' => 'User with the given email address not found'
```
Comment (sahajmalla) [inline app/Http/Requests/ForgotPasswordRequest.php:34]: is this necessary?

File: app/Http/Requests/PasswordResetRequest.php [hunk 4]
```diff
@@ -24,9 +28,25 @@ public function authorize()
...
+                    ->symbols(),
+                'confirmed',
+                new CheckHashedPasswordValue($this->email)
+            ],
+        ];
+    }
+
+    public function messages()
+    {
+        return [
```
Comment (sahajmalla) [inline app/Http/Requests/PasswordResetRequest.php:48]: is this necessary?

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Http/Requests/PasswordResetRequest.php (hunk 4 above)
Comment (piyushmhrzn) [reply app/Http/Requests/PasswordResetRequest.php:48]: Laravel Default Message was "The selected email is invalid."
So, I used custom message "User with the given email address not found" for more specific message.
Is this specific message justified here?

Comment (sahajmalla) [reply app/Http/Requests/PasswordResetRequest.php:48]: yes that is justified.

File: app/Http/Controllers/Actions/Auth/ForgotPasswordAction.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $validatedForgetPass = $request->validated();
+        try {
+            $status = $this->userService->forgotPassword($validatedForgetPass);
+            if(!$status) {
+                return $this->errorResponse([], "Failed to send email. Please try again later."); 
+            }
+            return $this->successResponse([], 'Reset email sent successfully', Response::HTTP_OK);
+
+        } catch (QueryException $queryException) {
```
Comment (sahajmalla) [inline app/Http/Controllers/Actions/Auth/ForgotPasswordAction.php:50]: why are you catching query exception here?

File: app/Http/Controllers/Actions/Auth/ResetPasswordAction.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        $validatedResetPass = $request->validated();
+        try {
+            // Check for a valid token or the valid email address of user
+            $status = $this->userService->resetPassword($validatedResetPass);
+            if (!$status){
+                return $this->errorResponse([], "Could not reset password. Please check your token or email address", Response::HTTP_FORBIDDEN);
+            }
+            return $this->successResponse([], 'Password reset successfully', Response::HTTP_OK);
+
+        }  catch (QueryException $queryException) {
```
Comment (sahajmalla) [inline app/Http/Controllers/Actions/Auth/ResetPasswordAction.php:49]: why are you catching query exception?

**Review by sahajmalla — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Actions/Auth/ForgotPasswordAction.php (hunk 5 above)
Comment (piyushmhrzn) [reply app/Http/Controllers/Actions/Auth/ForgotPasswordAction.php:50]: Is it already handled by forgot password request when we use exists for email?

**Review by sahajmalla — APPROVED** [review]

**Pull request merged**